File: app/src/main/java/com/bank/actionmode/DragExperimentTouchListener.java
package com.bank.actionmode;

import android.animation.ObjectAnimator;
import android.content.res.Resources;
import android.util.DisplayMetrics;
import android.util.Log;
import android.util.TypedValue;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.BounceInterpolator;

public class DragExperimentTouchListener implements View.OnTouchListener {

    private float dragDistance;
    boolean isOpen = false;

    public DragExperimentTouchListener(float initalX, float initialY, View view, View main) {
        lastX = initalX;
        lastY = initialY;
        this.view = view;
        this.main = main;
    }

    View view = null;
    View main = null;
    boolean isDragging = false;
    float lastX;
    float lastY;
    float deltaX;
    float deltaY;

    @Override
    public boolean onTouch(View view, MotionEvent event) {
        int action = event.getAction();
        if (action == MotionEvent.ACTION_DOWN && !isDragging) {
            isDragging = true;
            if (isOpen) deltaX = event.getX()-dpToPx(80);
             else deltaX = event.getX();
            return false;
        } else if (isDragging) {
            if (action == MotionEvent.ACTION_MOVE) {
                dragDistance = deltaX- event.getX();
                Log.d("distance", dragDistance+"");
                if (dragDistance <= dpToPx(80f) && dragDistance > 0) {
                    this.view.setTranslationX(-dragDistance);
                    main.setTranslationX((-dragDistance / 2) - (dpToPx(-40)));
                }

                if (dragDistance >= dpToPx(-80f) && dragDistance < 0) {

                    int width = main.getResources().getDisplayMetrics().widthPixels;

                    this.view.setTranslationX(-(width-dragDistance));
                  //  main.setTranslationX((-(width-dragDistance) / 2) - (dpToPx(-40)));
                }


                return false;
            } else if (action == MotionEvent.ACTION_UP) {
                isDragging = false;
                lastX = event.getX();
                dragDistance = deltaX-lastX;

                if (dragDistance>0 && Math.abs(dragDistance) > dpToPx(60))
                    open();
                else
                   if (!isOpen) close();

                if (dragDistance<=0 && Math.abs(dragDistance) > dpToPx(30))
                    close();
                else
                   if (isOpen) open();

                return false;
            } else if (action == MotionEvent.ACTION_CANCEL) {



                dragDistance = deltaX-lastX;
                if (dragDistance>0 && Math.abs(dragDistance) > dpToPx(60))
                    open();
                else
                if (!isOpen)  close();

                if (dragDistance<=0 && Math.abs(dragDistance) > dpToPx(30))
                    close();
                else
                if (isOpen)  open();

                lastX = event.getX();
                isDragging = false;
                return true;
            }
        }

        return false;
    }

    void open() {
        float x = -dpToPx(80);
        ObjectAnimator animator = ObjectAnimator.ofFloat(view, "translationX", x);
        animator.setDuration(700);
        animator.addUpdateListener(animation -> {
                    view.setTranslationX(((Float) (animation.getAnimatedValue())));
                    main.setTranslationX(((Float) (animation.getAnimatedValue()) / 2) + dpToPx(40f));
                }
        );
        animator.setInterpolator(new BounceInterpolator());
        animator.start();
        isOpen = true;
    }

    void close() {
        float x = 0f;
        ObjectAnimator animator = ObjectAnimator.ofFloat(view, "translationX", x);
        animator.setDuration(700);
        animator.addUpdateListener(animation -> {
                    view.setTranslationX(((Float) (animation.getAnimatedValue())));
                    main.setTranslationX(((Float) (animation.getAnimatedValue()) / 2) + dpToPx(40f));
                }
        );
        animator.setInterpolator(new BounceInterpolator());
        animator.start();

        isOpen = false;
    }

    float dpToPx(float d) {
        Resources r = main.getResources();
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, d, r.getDisplayMetrics());
    }

    float pxToDp(float p) {

        return p / ((float) main.getResources().getDisplayMetrics().densityDpi / DisplayMetrics.DENSITY_DEFAULT);
    }

}

File: app/src/main/java/com/bank/actionmode/DragExperimentTouchListener.kt
package com.bank.actionmode

import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.util.Log
import android.util.TypedValue
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import android.view.ViewParent
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.animation.AccelerateInterpolator
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.roundToLong

class DragExperimentTouchListener(var parent: ViewParent, private var lastX: Float, view: View?, main: View?) :
    OnTouchListener {
    private var dragDistanceX = 0f
    private var dragDistanceY = 0f
    private var isOpen = false
    private var view: View? = null
    private var main: View? = null
    private var isDragging = false
    private var deltaX = 0f
    private var deltaY = 0f
    private var ratio = 0f

    private var dontPass = true

    init {
        this.view = view
        this.main = main
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouch(view: View, event: MotionEvent): Boolean {
        val action = event.action
        if (action == MotionEvent.ACTION_DOWN && !isDragging) {
            parent.requestDisallowInterceptTouchEvent(false)
            isDragging = true
            deltaX = if (isOpen) event.x - dpToPx(80f) else event.x
            deltaY = if (isOpen) event.y - dpToPx(80f) else event.y
            return false
        } else if (isDragging) {
            when (action) {
                MotionEvent.ACTION_MOVE -> {
                    this.dragDistanceX = deltaX - event.x
                    this.dragDistanceY = deltaY - event.y

                    //todo
                    if (abs(Math.toDegrees(atan2(dragDistanceY, dragDistanceX).toDouble())) !in 80f..100f) {
                        parent.requestDisallowInterceptTouchEvent(true)
                    }

                    if (!isOpen)
                        if (dragDistanceX <= dpToPx(80f) && dragDistanceX > 0) {
                            this.view!!.translationX = -dragDistanceX
                            main!!.translationX = -dragDistanceX / 2 - dpToPx(-40f)
                            //todo
                            ratio = abs(dragDistanceX / dpToPx(80f))
                        }

                    if (isOpen) {
                        dragDistanceX += dpToPx(80f)
                        if (dragDistanceX >= dpToPx(-80f) && dragDistanceX < 0) {
                            this.view!!.translationX = dpToPx(-80f) - dragDistanceX
                            main!!.translationX = (dpToPx(-80f) - dragDistanceX) / 2 - dpToPx(-40f)
                            //todo
                            ratio = abs((dpToPx(-80f) - dragDistanceX) / dpToPx(80f))
                        }
                    }

                    return false
                }
                MotionEvent.ACTION_UP -> {
                    dontPass = true
                    isDragging = false
                    lastX = event.x
                    dragDistanceX = deltaX - lastX
                    if (dragDistanceX > 0 && abs(dragDistanceX) > dpToPx(60f)) open() else if (!isOpen) close()
                    if (dragDistanceX <= 0 && abs(dragDistanceX) > dpToPx(60f)) close() else if (isOpen) open()
                    return false
                }
                MotionEvent.ACTION_CANCEL -> {
                    dontPass = true
                    dragDistanceX = deltaX - lastX
                    if (dragDistanceX > 0 && abs(dragDistanceX) > dpToPx(60f)) open() else if (!isOpen) close()
                    if (dragDistanceX <= 0 && abs(dragDistanceX) > dpToPx(60f)) close() else if (isOpen) open()
                    lastX = event.x
                    isDragging = false
                    return true
                }
            }
        }
        return false
    }

    @SuppressLint("ObjectAnimatorBinding")
    fun open() {
        Log.d("note ratio", "open: $ratio")
        val x = -dpToPx(80f)
        val animator = ObjectAnimator.ofFloat(view, "translationX", x)
        //todo
        animator.duration = (400*ratio).roundToLong()
        animator.addUpdateListener { animation: ValueAnimator ->
            view!!.translationX = (animation.animatedValue as Float)
            main!!.translationX = animation.animatedValue as Float / 2 + dpToPx(40f)
        }
        animator.interpolator = AccelerateDecelerateInterpolator()
        animator.start()
        isOpen = true
    }

    @SuppressLint("ObjectAnimatorBinding")
    fun close() {
        Log.d("note ratio", "close: $ratio")
        val x = 0f
        val animator = ObjectAnimator.ofFloat(view, "translationX", x)
        //todo
        animator.duration = (400*ratio).roundToLong()
        animator.addUpdateListener { animation: ValueAnimator ->
            view!!.translationX = (animation.animatedValue as Float)
            main!!.translationX = animation.animatedValue as Float / 2 + dpToPx(40f)
        }
        animator.interpolator = AccelerateDecelerateInterpolator()
        animator.start()
        isOpen = false
    }

    private fun dpToPx(d: Float): Float {
        val r = main!!.resources
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, d, r.displayMetrics)
    }

}